File: src/main/java/at/yawk/fiction/Author.java
package at.yawk.fiction;

import at.yawk.fiction.impl.ao3.Ao3Author;
import at.yawk.fiction.impl.fanfiction.FfnAuthor;
import at.yawk.fiction.impl.fimfiction.FimAuthor;
import com.fasterxml.jackson.annotation.JsonSubTypes;
import com.fasterxml.jackson.annotation.JsonTypeInfo;
import lombok.Data;

/**
 * @author yawkat
 */
@Data
@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS)
@JsonSubTypes({
        @JsonSubTypes.Type(FfnAuthor.class),
        @JsonSubTypes.Type(FimAuthor.class),
        @JsonSubTypes.Type(Ao3Author.class),
})
public abstract class Author {
    String name;
}

File: src/main/kotlin/at/yawk/fiction/Author.kt
package at.yawk.fiction

import com.fasterxml.jackson.annotation.JsonTypeInfo

/**
 * @author yawkat
 */
data class Author(
        val name: String? = null,
        val providerData: AuthorProviderData
) : Mergeable<Author> {
    override fun merge(other: Author): Author {
        return Author(
                name = name ?: other.name,
                providerData = providerData.merge(other.providerData)
        )
    }
}

@JsonTypeInfo(use = JsonTypeInfo.Id.MINIMAL_CLASS)
interface AuthorProviderData : Mergeable<AuthorProviderData>, ProviderData